File: packages/forge-canvas-dart/lib/src/canvas_registry.dart
import 'package:flutter/widgets.dart';

import 'lint.dart';

/// One canvas component — a name, a Flutter widget builder, and an
/// optional JSON schema for its props.
class CanvasComponent {
  final String name;
  final Widget Function(Map<String, dynamic> props) builder;
  final Map<String, dynamic>? propsSchema;

  const CanvasComponent({
    required this.name,
    required this.builder,
    this.propsSchema,
  });
}

/// Resolution result: the matched component plus any lint issues found
/// against the provided props (empty in release mode).
class CanvasResolution {
  final CanvasComponent entry;
  final List<LintIssue> issues;

  const CanvasResolution({required this.entry, required this.issues});
}

/// Resolves component names (from backend payloads) to Flutter widget
/// builders. Mirrors the Vue/Svelte registry shape.
class CanvasRegistry {
  final Map<String, CanvasComponent> _entries = {};

  CanvasRegistry([List<CanvasComponent> initial = const []]) {
    for (final e in initial) {
      register(e);
    }
  }

  void register(CanvasComponent entry) {
    if (_entries.containsKey(entry.name)) {
      throw StateError('canvas component "${entry.name}" is already registered');
    }
    _entries[entry.name] = entry;
  }

  CanvasComponent? resolve(String name) => _entries[name];

  /// Resolve + validate props against the component's schema.
  CanvasResolution? lintAndResolve(String name, Map<String, dynamic> props) {
    final entry = _entries[name];
    if (entry == null) return null;
    final issues = lintProps(entry.propsSchema, props);
    warnOnLintIssues(entry.name, issues);
    return CanvasResolution(entry: entry, issues: issues);
  }

  Iterable<CanvasComponent> entries() => _entries.values;
}

File: packages/canvas-svelte/src/canvas-registry.ts
// Canvas component registry — Svelte 5 variant.

import type { Component } from 'svelte'

import { lintProps, warnOnLintIssues } from './lint'

export interface CanvasComponent<Props = Record<string, unknown>> {
  name: string
  component: Component
  propsSchema?: Record<string, unknown>
}

export interface CanvasResolution<Props = Record<string, unknown>> {
  entry: CanvasComponent<Props>
  issues: readonly { field: string; message: string }[]
}

export interface CanvasRegistry {
  register(entry: CanvasComponent): void
  resolve(name: string): CanvasComponent | null
  lintAndResolve(
    name: string,
    props: Record<string, unknown>,
  ): CanvasResolution | null
  entries(): readonly CanvasComponent[]
}

export function createCanvasRegistry(initial: CanvasComponent[] = []): CanvasRegistry {
  const entries = new Map<string, CanvasComponent>()
  for (const e of initial) entries.set(e.name, e)

  return {
    register(entry) {
      if (entries.has(entry.name)) {
        throw new Error(`canvas component "${entry.name}" is already registered`)
      }
      entries.set(entry.name, entry)
    },
    resolve(name) {
      return entries.get(name) ?? null
    },
    lintAndResolve(name, props) {
      const entry = entries.get(name)
      if (!entry) return null
      const issues = lintProps(entry.propsSchema, props)
      warnOnLintIssues(entry.name, issues)
      return { entry, issues }
    },
    entries() {
      return Array.from(entries.values())
    },
  }
}
